### ISSUE TITLE
CUDA: Atomics tests fail with NumPy 1.20

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#6965]
**EDIT**: Originally this was reported as an issue with Python 3.9, but it actually seems to be NumPy 1.20 which causes the issue.

Tested with Numba 0.53.1 and Python 3.9.4.

```
$ numba -s
System info:
--------------------------------------------------------------------------------
__Time Stamp__
Report started (local time)                   : 2021-04-23 10:50:10.994280
UTC start time                                : 2021-04-23 09:50:10.994284
Running time (s)                              : 0.872906

__Hardware Information__
Machine                                       : x86_64
CPU Name                                      : skylake-avx512
CPU Count                                     : 12
Number of accessible CPUs                     : 12
List of accessible CPUs cores                 : 0-11
CFS Restrictions (CPUs worth of runtime)      : None

CPU Features                                  : 64bit adx aes avx avx2 avx512bw
                                                avx512cd avx512dq avx512f avx512vl
                                                bmi bmi2 clflushopt clwb cmov cx16
                                                cx8 f16c fma fsgsbase fxsr invpcid
                                                lzcnt mmx movbe pclmul pku popcnt
                                                prfchw rdrnd rdseed rtm sahf sse
                                                sse2 sse3 sse4.1 sse4.2 ssse3
                                                xsave xsavec xsaveopt xsaves

Memory Total (MB)                             : 95270
Memory Available (MB)                         : 91775

__OS Information__
Platform Name                                 : Linux-5.4.0-72-generic-x86_64-with-glibc2.27
Platform Release                              : 5.4.0-72-generic
OS Name                                       : Linux
OS Version                                    : #80~18.04.1-Ubuntu SMP Mon Apr 12 23:26:25 UTC 2021
OS Specific Version                           : ?
Libc Version                                  : glibc 2.27

__Python Information__
Python Compiler                               : GCC 7.3.0
Python Implementation                         : CPython
Python Version                                : 3.9.4
Python Locale                                 : en_GB.UTF-8

__LLVM Information__
LLVM Version                                  : 10.0.1

__CUDA Information__
CUDA Device Initialized                       : True
CUDA Driver Version                           : 11030
CUDA Detect Output:
Found 2 CUDA devices
id 0    b'NVIDIA Quadro RTX 8000'                              [SUPPORTED]
                      compute capability: 7.5
                           pci device id: 0
                              pci bus id: 21
id 1    b'NVIDIA Quadro RTX 8000'                              [SUPPORTED]
                      compute capability: 7.5
                           pci device id: 0
                              pci bus id: 45
Summary:
        2/2 devices are supported

CUDA Libraries Test Output:
Finding cublas from System
        named  libcublas.so.11.2.2.29
        trying to open library...       ok
Finding cusparse from System
        named  libcusparse.so.11.3.1.29
        trying to open library...       ok
Finding cufft from System
        named  libcufft.so.10.4.0.29
        trying to open library...       ok
Finding curand from System
        named  libcurand.so.10.2.3.29
        trying to open library...       ok
Finding nvvm from System
        named  libnvvm.so.4.0.0
        trying to open library...       ok
Finding cudart from System
        named  libcudart.so.11.2.29
        trying to open library...       ok
Finding cudadevrt from System
        named  libcudadevrt.a
Finding libdevice from System
        searching for compute_20...     ok
        searching for compute_30...     ok
        searching for compute_35...     ok
        searching for compute_50...     ok


__ROC information__
ROC Available                                 : False
ROC Toolchains                                : None
HSA Agents Count                              : 0
HSA Agents:
None
HSA Discrete GPUs Count                       : 0
HSA Discrete GPUs                             : None

__SVML Information__
SVML State, config.USING_SVML                 : False
SVML Library Loaded                           : False
llvmlite Using SVML Patched LLVM              : True
SVML Operational                              : False

__Threading Layer Information__
TBB Threading Layer Available                 : True
+-->TBB imported successfully.
OpenMP Threading Layer Available              : True
+-->Vendor: GNU
Workqueue Threading Layer Available           : True
+-->Workqueue imported successfully.

__Numba Environment Variable Information__
None found.

__Conda Information__
Conda Build                                   : 3.20.1
Conda Env                                     : 4.9.2
Conda Platform                                : linux-64
Conda Python Version                          : 3.8.3.final.0
Conda Root Writable                           : True

__Installed Packages__
_libgcc_mutex             0.1                        main
blas                      1.0                         mkl
ca-certificates           2021.4.13            h06a4308_1
certifi                   2020.12.5        py39h06a4308_0
intel-openmp              2021.2.0           h06a4308_610
ld_impl_linux-64          2.33.1               h53a641e_7
libffi                    3.3                  he6710b0_2
libgcc-ng                 9.1.0                hdf63c60_0
libllvm10                 10.0.1               hbcb73fb_5
libstdcxx-ng              9.1.0                hdf63c60_0
llvmlite                  0.36.0           py39h612dafd_4
mkl                       2021.2.0           h06a4308_296
mkl-service               2.3.0            py39h27cfd23_1
mkl_fft                   1.3.0            py39h42c9631_2
mkl_random                1.2.1            py39ha9443f7_2
ncurses                   6.2                  he6710b0_1
numba                     0.53.1           py39ha9443f7_0
numpy                     1.20.1           py39h93e21f0_0
numpy-base                1.20.1           py39h7d8b39e_0
openssl                   1.1.1k               h27cfd23_0
pip                       21.0.1           py39h06a4308_0
python                    3.9.4                hdb3f193_0
readline                  8.1                  h27cfd23_0
setuptools                52.0.0           py39h06a4308_0
six                       1.15.0           py39h06a4308_0
sqlite                    3.35.4               hdfb4753_0
tbb                       2020.3               hfd86e86_0
tk                        8.6.10               hbc83047_0
tzdata                    2020f                h52ac0ba_0
wheel                     0.36.2             pyhd3eb1b0_0
xz                        5.2.5                h7b6447c_0
zlib                      1.2.11               h7b6447c_3

No errors reported.
```

Test errors:

```
======================================================================
ERROR: test_atomic_nanmax_int32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1279, in test_atomic_nanmax_int32
    self.check_atomic_nanmax(dtype=np.int32, lo=-65535, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1271, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_int64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1286, in test_atomic_nanmax_int64
    self.check_atomic_nanmax(dtype=np.int64, lo=-65535, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1271, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_uint32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1282, in test_atomic_nanmax_uint32
    self.check_atomic_nanmax(dtype=np.uint32, lo=0, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1271, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_uint64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1290, in test_atomic_nanmax_uint64
    self.check_atomic_nanmax(dtype=np.uint64, lo=0, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1271, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_int32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1332, in test_atomic_nanmin_int32
    self.check_atomic_nanmin(dtype=np.int32, lo=-65535, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1323, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_int64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1339, in test_atomic_nanmin_int64
    self.check_atomic_nanmin(dtype=np.int64, lo=-65535, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1323, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_uint32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1335, in test_atomic_nanmin_uint32
    self.check_atomic_nanmin(dtype=np.uint32, lo=0, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1323, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_uint64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1343, in test_atomic_nanmin_uint64
    self.check_atomic_nanmin(dtype=np.uint64, lo=0, hi=65535)
  File "/home/gmarkall/miniconda3/envs/NumbaGPUTest/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_atomics.py", line 1323, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

----------------------------------------------------------------------
Ran 1124 tests in 138.542s

FAILED (errors=8, skipped=11, expected failures=1)
```

**Comment by gmarkall** [comment]
The NumPy version is what determines whether there is an issue. With NumPy 1.20.1:

```
$ python -m numba.runtests numba.cuda.tests.cudapy.test_atomics
.../home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
......................................................................EE....EE....EE....EE......................
======================================================================
ERROR: test_atomic_nanmax_int32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1388, in test_atomic_nanmax_int32
    self.check_atomic_nanmax(dtype=np.int32, lo=-65535, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1380, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_int64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1395, in test_atomic_nanmax_int64
    self.check_atomic_nanmax(dtype=np.int64, lo=-65535, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1380, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_uint32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1391, in test_atomic_nanmax_uint32
    self.check_atomic_nanmax(dtype=np.uint32, lo=0, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1380, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmax_uint64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1399, in test_atomic_nanmax_uint64
    self.check_atomic_nanmax(dtype=np.uint64, lo=0, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1380, in check_atomic_nanmax
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_int32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1441, in test_atomic_nanmin_int32
    self.check_atomic_nanmin(dtype=np.int32, lo=-65535, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1432, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_int64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1448, in test_atomic_nanmin_int64
    self.check_atomic_nanmin(dtype=np.int64, lo=-65535, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1432, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_uint32 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1444, in test_atomic_nanmin_uint32
    self.check_atomic_nanmin(dtype=np.uint32, lo=0, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1432, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

======================================================================
ERROR: test_atomic_nanmin_uint64 (numba.cuda.tests.cudapy.test_atomics.TestCudaAtomics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1452, in test_atomic_nanmin_uint64
    self.check_atomic_nanmin(dtype=np.uint64, lo=0, hi=65535)
  File "/home/gmarkall/numbadev/numba/numba/cuda/tests/cudapy/test_atomics.py", line 1432, in check_atomic_nanmin
    vals[1::2] = np.nan
ValueError: cannot convert float NaN to integer

----------------------------------------------------------------------
Ran 120 tests in 9.740s

FAILED (errors=8)
```

With NumPy 1.19.2:

```
$ python -m numba.runtests numba.cuda.tests.cudapy.test_atomics
.../home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
./home/gmarkall/numbadev/numba/numba/cuda/compiler.py:1150: NumbaDeprecationWarning: inspect_asm will always return a dict in future
  warn('inspect_asm will always return a dict in future',
................................................................................................................
----------------------------------------------------------------------
Ran 120 tests in 9.276s

OK
```

**Comment by esc** [comment]
Datashader showed up with this in the numba-integration-testing suite too:

https://app.circleci.com/pipelines/github/numba/numba-integration-testing/641/workflows/f8fed781-8b5f-485a-accd-fc81ab3c5e0c/jobs/7062

**Comment by gmarkall** [comment]
Fixed in #6995.
